kroxylicious-sample/src/test/java/io/kroxylicious/sample/SampleFetchResponseFilterTest.java

```diff
@@ -154,15 +184,60 @@ private static FetchResponseData buildFetchResponseData(String transformValue) {
     }
 
     private static List<String> unpackFetchResponseData(FetchResponseData responseData) {
-        var responses = responseData.responses();
         var values = new ArrayList<String>();
-        for (FetchResponseData.FetchableTopicResponse response : responses) {
+        var records = unpackRecords(responseData);
+        assertThat(records).isNotNull();
+        for (Record record : records) {
+            values.add(new String(StandardCharsets.UTF_8.decode(record.value()).array()));
+        }
+        return values;
+    }
+
+    private static List<Record> unpackRecords(FetchResponseData responseData) {
+        List<Record> records = new ArrayList<>();
+        for (FetchResponseData.FetchableTopicResponse response : responseData.responses()) {
             for (FetchResponseData.PartitionData partitionData : response.partitions()) {
-                for (Record record : ((MemoryRecords) partitionData.records()).records()) {
-                    values.add(new String(StandardCharsets.UTF_8.decode(record.value()).array()));
+                for (Record record : ((Records) partitionData.records()).records()) {
+                    records.add(record);
                 }
             }
         }
-        return values;
+        return records;
+    }
+
+    private static void compareRecords(FetchResponseData a, FetchResponseData b) {
+        var aRecords = unpackRecords(a).listIterator();
+        var bRecords = unpackRecords(b).listIterator();
+        assertThat(aRecords).isNotNull();
+        assertThat(bRecords).isNotNull();
+        while (aRecords.hasNext()) {
+            assertThat(bRecords).hasNext();
+            checkRecordMetadataMatches(aRecords.next(), bRecords.next());
+        }
+    }
+
+    private static void checkRecordMetadataMatches(Record a, Record b) {
+        // We don't compare record size, value, or value size as we expect the filter to change these
+        assertThat(a.offset()).isEqualTo(b.offset());
+        assertThat(a.sequence()).isEqualTo(b.sequence());
+        assertThat(a.timestamp()).isEqualTo(b.timestamp());
+        assertThat(a.keySize()).isEqualTo(b.keySize());
+        assertThat(a.hasKey()).isEqualTo(b.hasKey());
+        assertThat(a.key()).isSameAs(b.key());
+        assertThat(a.hasValue()).isEqualTo(b.hasValue());
+        assertThat(a.isCompressed()).isEqualTo(b.isCompressed());
```

Comment:
The above would give a failure message along the lines of 
```
java.lang.AssertionError: 
Expecting actual:
  DefaultRecord(offset=0, timestamp=-1, key=0 bytes, value=45 bytes)
to be equal to:
  DefaultRecord(offset=11, timestamp=1692577439343, key=0 bytes, value=5 bytes)
when recursively comparing field by field, but found the following 4 differences:

field/property 'headers' differ:
- actual value  : [RecordHeader(key = myKey, value = [109, 121, 86, 97, 108, 117, 101])]
- expected value: []
actual and expected values are arrays of different size, actual size=1 when expected size=0

field/property 'offset' differ:
- actual value  : 0L
- expected value: 11L

field/property 'sequence' differ:
- actual value  : -1
- expected value: 10

field/property 'timestamp' differ:
- actual value  : -1L
- expected value: 1692577439343L

The recursive comparison was performed with this configuration:
- the following fields were ignored in the comparison: sizeInBytes, valueSize, value
- no overridden equals methods were used in the comparison (except for java types)
- these types were compared with the following comparators:
  - java.lang.Double -> DoubleComparator[precision=1.0E-15]
  - java.lang.Float -> FloatComparator[precision=1.0E-6]
  - java.nio.file.Path -> lexicographic comparator (Path natural order)
- actual and expected objects and their fields were compared field by field recursively even if they were not of the same type, this allows for example to compare a Person to a PersonDto (call strictTypeChecking(true) to change that behavior).
- the introspection strategy used was: DefaultRecursiveComparisonIntrospectionStrategy
```

Reply:
That shortens things considerably. It also means there's no real need to split out the header check into its own method, it can live with the rest of the metadata checks. I've made that change in both sets of tests now.